src/main/webapp/src/sessions/gapi.session.ts

```diff
@@ -0,0 +1,217 @@
+
+const GAPI_CLIENT_ID = CLIENT_ID;
+const API_KEY = CALENDAR_API_KEY;
+
+const DISCOVERY_DOCS = [
+  'https://www.googleapis.com/discovery/v1/apis/drive/v3/rest',
+];
+const SCOPES = 'https://www.googleapis.com/auth/calendar';
+
+declare const gapi: any;
+
+@Injectable()
+export class GapiSession {
+  public auth2: any;
+  hasConsent = false;
+  tasks = new TaskComponent();
+
+  constructor(private zone: NgZone) {}
+
+  ngOnInit() {
+    this.loadClient();
+  }
+
+  /**
+   * Load the Calendar API client and provide authentication
+   */
+  loadClient(): void {
+    gapi.load('client:auth2', () => {
+      gapi.auth2
+        .init({
+          apiKey: API_KEY,
+          clientId: GAPI_CLIENT_ID,
+          discoveryDocs: DISCOVERY_DOCS,
+          scope: SCOPES,
+        })
+        .then(auth => {
+          this.zone.run(() => {
+            this.auth2 = auth;
+          });
+        });
+    });
+  }
+
+  /**
+   * Returns whether user is signed in with Google
+   */
+  get isSignedIn(): boolean {
+    return this.auth2.isSignedIn.get();
+  }
+
+  /**
+   * Returns whether the user has accepted the consent prompt in the current session
+   */
+  get consent(): boolean {
+    return this.hasConsent;
+  }
+
+  /**
+   * Google Authentication; displays consent prompt for Google Calendar API
+   */
+  signIn() {
+    return this.auth2
+      .signIn({
+        prompt: 'consent',
+      })
+      .then(() => {
+        this.hasConsent = true;
+      });
+  }
+
+  /**
+   * Sign out from Google Auth
+   */
+  signOut(): void {
+    this.auth2.signOut();
+  }
+
+  /**
+   * Retrieve calendar events given a date and call function to display the events in the page
+   * @param selectedDate - start date and time (min)
+   * @param selectedDateMax - end of the day of selected date (max)
+   */
+  listEvents(
+    gardenName: string,
+    selectedDate: string,
+    selectedDateMax: string
+  ) {
+    gapi.client.load('calendar', 'v3', () => {
+      gapi.client.calendar.events
+        .list({
+          calendarId: 'primary',
+          timeMin: selectedDate,
+          timeMax: selectedDateMax,
+          singleEvents: true,
+          orderBy: 'startTime',
+        })
+        .then(response => {
+          const events = response.result.items;
+          this.tasks.clearTasks();
+          if (events.length > 0) {
+            for (let i = 0; i < events.length; i++) {
+              const event = events[i];
+              if (event.summary.indexOf(gardenName) !== -1) {
+                const eventTime = this.getEventTime(event.start.dateTime);
+                this.tasks.createTaskElement(
+                  eventTime,
+                  event.summary,
+                  event.attendees,
+                  event.description
+                );
+              }
+            }
+          }
+        });
+    });
+  }
+
+  /**
+   * Get the start time of the event otherwise return an All Day string
```

Comment:
nit: it would be nice if you include the expected format of the `dateTime` string. Also, I believe the correct type for dateTime would be `string | undefined` since it may be omitted.